pageserver/src/tenant/timeline.rs

```diff
@@ -1450,7 +1459,7 @@ impl Timeline {
                     reconstruct_state,
                 )?;
                 cont_lsn = lsn_floor;
-                traversal_path.push((result, cont_lsn, layer));
+                traversal_path.push((result, cont_lsn, layer.as_pure_layer()));
```

Comment:
This is current the only place we use this method.
Also, `traversal_path` is only passed into `layer_traversal_error` function which then only calls `l.short_id()` for some message creation.

If my assessments are correct, I would remove the entire upcast logic with `fn as_pure_layer<'a>(self: Arc<Self>) -> Arc<dyn PureLayer>` and its impls, and would do a small conversion before.

Reply:
What conversion are you referring to? So that `traversal_path` only contains `short_id()`?

Reply:
Yes, just basically call `short_id()` before sending `traversal_path` into `layer_traversal_error`: this way, we can avoid knowing about traits in that function at all.

Reply:
Will introduce the concept of a `TraversalId` in the next push.